[WIP] Message Index

vyzo opened the pull request:
See #10326 

## Design Overview
This departs from the discussion in #10326 by implementing a simpler solution that has no memory component and stores everything in sqlite.
On head change notifications, we first delete data from reverted tipsets and then insert data for applied tipsets (in a transaction).
This allows us to trivially deal with finality commitments.

raulk commented on chain/index/interface.go:
```diff
@@ -0,0 +1,31 @@
+	Epoch abi.ChainEpoch
+	// the tipset where this messages executed
+	Tipset types.TipSetKey
+	// the first block in the tipset where the message was executed
+	Block cid.Cid
+	// the index of the message in the block
+	Index int
```
I think you can remove the Block and instead have the canonicalized execution index (which is the index at which the receipt will be stored in the AMT).

vyzo replied:
ok, I will see whether I can get rid of it when I get to implementation details.

vyzo replied:
Yeah, we don't need it; dropped it from the interface.

raulk commented on chain/index/interface.go:
```diff
@@ -0,0 +1,31 @@
+	// the epoch whre this message was executed
+	Epoch abi.ChainEpoch
+	// the tipset where this messages executed
+	Tipset types.TipSetKey
```
Let's use a TipsetCID instead. The TipSetKey is wasteful. It will imply an extra lookup to go from TipsetCid to TipsetKey, but it's worth it to save an average of 4x times the amount of bytes that would otherwise go here, on every record. Discarding a second table + foreign key + joins because that's too much complexity and antithetical to the main goal of speeding things up.

vyzo replied:
done.

raulk commented on chain/index/msgindex.go:
```diff
@@ -0,0 +1,433 @@
+		return nil, xerrors.Errorf("error opening msgindex database: %w", err)
+	}
+
+	if mkdb {
+		err = createTables(db)
+		if err != nil {
+			return nil, xerrors.Errorf("error creating msgindex database: %w", err)
+		}
```
Can save all this dance by using `CREATE TABLE IF NOT EXISTS`. See https://github.com/filecoin-project/lotus/blob/master/chain/events/filter/index.go#L32-L61.

vyzo replied:
we can, but we still need to dance for reconciliation.

raulk replied:
Aren't index initialization and reconciliation independent of one another?

vyzo replied:
yeah, all we need to do is stat to know if it already exists.
Removed the unnecessary parts of the dance.

raulk commented on chain/index/msgindex.go:
```diff
@@ -0,0 +1,433 @@
+		}
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
```
This isn't the right context to use. Let's make the constructor accept a context, like the [Mpool](https://github.com/filecoin-project/lotus/blob/58900a70333a11a903cf9fe3f29e6a5c309cb000/chain/messagepool/messagepool.go#L371), and feed the lifecycle context during dependency injection.

vyzo replied:
I disagree in this. This is a background component, it is certainly the right context.

Passing a lifecycle context is an antipattern, and we are doing it wrong.
I am sure @stebalien has strong opinions here.

raulk replied:
We shouldn't be using contexts for lifecycle management, agree, but that's Lotus' reality today and we should just adhere to it. If we want to move away from that pattern, it should be done as part of a larger refactor, not in a piecemeal way.

vyzo replied:
I will invoke the @Stebalien shell here. Codewise it is easy to hang this context from the lifecycle context.

magik6k replied:
The lifecycle context should be fine to use. If it is created before the OnClose hook is set, it is guaranteed to be cancelled after Close is called - which is the whole point of the lifecycle system. Lifecycle contexts are somewhat of a hack, but in most cases they just make cleanup of things slightly more robust.

The other thing that context is used for is metrics / tracing, and in this specific case I believe there actually are metrics which will be dropped due to this background context not being wired to the main MetricsCtx

vyzo replied:
well ok, i will wire it then.

vyzo replied:
done.

raulk commented on chain/index/msgindex.go:
```diff
@@ -0,0 +1,433 @@
+
+	err = msgIndex.prepareStatements()
+	if err != nil {
+		err2 := db.Close()
+		if err2 != nil {
```
```suggestion
		if err := db.Close(); err != nil {
```

vyzo replied:
sure.

raulk commented on chain/index/msgindex.go:
```diff
@@ -0,0 +1,433 @@
+// init utilities
+func createTables(db *sql.DB) error {
+	// Just a single table for now; ghetto, but this an index so we denormalize to avoid joins.
+	if _, err := db.Exec("CREATE TABLE Messages (cid VARCHAR(80) PRIMARY KEY, tipset VARCHAR(80), xepoch INTEGER, xindex INTEGER)"); err != nil {
```
Can we please use the same pattern for initializing tables as the event index? https://github.com/filecoin-project/lotus/blob/master/chain/events/filter/index.go#L32-L61

- Have the DDLs hoisted as a global var for better visiblity.
- Use CREATE IF NOT EXISTS
- Add a version table to facilitate future migrations.
- Lower case table name
- `cid` should be of BLOB type?
- same for `tipset`
- `exec_epoch` and `exec_index` please -- costs nothing and it's immediately obvious.

vyzo replied:
Ok.

Cid shouldnt be blib if we want to primary key it imho.

vyzo replied:
Same for tipset if we use an index, which we will.

vyzo replied:
Done:
- hoisted DDLs and stmts on top to vars and consts
- used CREATE IF NOT EXISTS
- added version table
- lower cased table name
- renamed tipset to tipset_cid, xepoch to just epoch, and dropped xindex.

See above for why I am keeping VARCHAR(80) for the cid types:
- we need primary key for cid
- we need index for tipset_cid

raulk commented on chain/index/msgindex.go:
```diff
@@ -0,0 +1,433 @@
+		return err
+	}
+
+	// TODO Should we add an index for tipset to speed up deletion on revert?
```
Yes, otherwise the cost of reverts will scale with the size of the DB, and that will be problematic.

vyzo replied:
yeah, makes sense, will add.

vyzo replied:
done.

raulk reviewed (commented):
Going through it, expect more comments shortly.

raulk commented on chain/index/msgindex.go:
```diff
@@ -0,0 +1,433 @@
+	// Invariant: after reconciliation, every tipset in the index is in the current chain; ie either
+	//  the chain head or reachable by walking the chain.
+	// Algorithm:
+	//  1. Count mesages in index; if none, trivially reconciled.
```
```suggestion
	//  1. Count messages in index; if none, trivially reconciled.
```

raulk commented on chain/index/msgindex.go:
```diff
@@ -0,0 +1,433 @@
+}
+
+func (x *msgIndex) prepareStatements() error {
+	stmt, err := x.db.Prepare("SELECT tipset, xepoch, xindex FROM Messages WHERE cid = ?")
+	if err != nil {
+		return xerrors.Errorf("prepare selectMsgStmt: %w", err)
+	}
+	x.selectMsgStmt = stmt
+
+	stmt, err = x.db.Prepare("INSERT INTO Messages VALUES (?, ?, ?, ?)")
+	if err != nil {
+		return xerrors.Errorf("prepare insertMsgStmt: %w", err)
+	}
+	x.insertMsgStmt = stmt
+
+	stmt, err = x.db.Prepare("DELETE FROM Messages WHERE tipset = ?")
+	if err != nil {
+		return xerrors.Errorf("prepare deleteTipSetStmt: %w", err)
+	}
+	x.deleteTipSetStmt = stmt
```
Let's hoist these as consts, makes them more visible.

See https://github.com/filecoin-project/lotus/blob/master/chain/events/filter/index.go#L65-L73.

vyzo replied:
done.

raulk commented on chain/index/msgindex.go:
```diff
@@ -0,0 +1,433 @@
+	}
+
+	for _, blk := range ts.Blocks() {
+		bmsgs, smsgs, err := x.cs.MessagesForBlock(ctx, blk)
```
This should use `MessagesForTipset` instead.

vyzo replied:
ack, will fix.

raulk commented on chain/index/msgindex.go:
```diff
@@ -0,0 +1,433 @@
+	}
+
+	tskey := tscid.String()
+	xepoch := int64(ts.Height())
```
This is not the execution epoch, but the _inclusion_ epoch.

vyzo replied:
fixed.

raulk commented on chain/index/msgindex.go:
```diff
@@ -0,0 +1,433 @@
+	for {
+		select {
+		case <-x.sema:
+			err := x.processHeadChanges(ctx)
```
Unfortunately this logic is not as straightforward as it appears here. This index only cares about executed tipsets, which means that we need to consume the parent of each head, but only when the head has actually been executed itself. Sadly this logic consumes speculative tipsets while they are being formed, which is not what we want.

In other words, we should track tipsets with a **confidence of 2**, i.e. once the **execution** tipset has been committed to chain. Then we can add the messages from the execution tipset, which is head.height - 2 at minimum, or more if there are null rounds.

```
inclusion ts <--- execution ts <--- head
```

---

Also, we should extract and reuse the [event observer](https://github.com/filecoin-project/lotus/blob/e65fae28de2a44947dd24af8c7dafcade29af1a4/chain/events/observer.go#L19) whcih already tracks tipsets, performs reverts, and does reconciliation by replaying a stream of reverts and applies for a particular chain path.

That component is actually not interdependent with the event index, so it can be extracted easily!

vyzo replied:
For readers of this pr: see comments below, we are going inclusion tipset.

raulk reviewed (commented):
More comments, let's reuse the event observer logic by extracting it first to a common component.

raulk commented on chain/stmgr/searchwait.go:
```diff
@@ -159,6 +173,21 @@ func (sm *StateManager) SearchForMessage(ctx context.Context, head *types.TipSet
+		return nil, nil, cid.Undef, err
+	}
+
+	r, foundMsg, err := sm.tipsetExecutedMessage(ctx, ts, mcid, m.VMMessage(), false)
+	return ts, r, foundMsg, err
+}
```
We should use the execution tipset and the execution index to just rescue the receipt and call it a day here, instead of handing off to `tipsetExecutedMessage`. However, @arajasek wanted to contain the changeset leaving out the execution index from the index, in which case this logic can stay.

vyzo replied:
we dropped the index from the index.

vyzo commented:
We discussed sync, and we think that:
1. It is ok to use the inclusion index.
2. We can drop the execution index.

In this light, we probably need to do very little. 

We can even keep the coaleacer instead of the event observer. Also note that confidence can be implemented at read time as well, by comparing inclusion height with the current head.

raulk reviewed (commented):
Just had a chat with @vyzo. While having the execution tipset and the execution index would be ideal, it'll be easier to get this merged if we stick to what the discussion in #10326 ends up suggesting. So:

1. Let's store the inclusion tipset, which is what this logic is actually storing now (even though the field is incorrectly named as _execution_ tipset -- fix that).
2. Let's remove the execution index.
3. Reuse the events observer, extracting it from the `chain/events` package since it's not coupled to events.
4. Set a confidence of 1 (see how [the `headEventsObserver` does it](https://github.com/filecoin-project/lotus/blob/e65fae28de2a44947dd24af8c7dafcade29af1a4/chain/events/events_height.go#L157)). Not 2 since we're tracking the inclusion tipset and not the execution tipset now.
5. Align the code with the existing event index to avoid code entropy.

## TODOs in addition to the above

(perhaps incomplete, @vyzo please add if I omitted something)

- [ ] Add itests.
- [ ] Add a backfill command in lotus-shed.
- [ ] Make sure we're using the right sqlite pragmas to: (1) fsync on transaction commit, and (2) allow to open the sqlite db from multiple processes (so that the backfill script can operate while the node is live).

vyzo commented:
Checklist looks good, but we might want to punt on it a bit if it is tricky.

It looks like the observer might be trivial to integrate to use it instead of the coalescer. 

A couple if additional considerations:
- backfill at startup the first time we create the index, for some epoch range. Not sure how good an idea it is, compared to shed, but thats why i wanr to punt on it in this pr.
- we may want to disable the fallback to prevent dos vectors. At the very least provide an option to limit the range may br necessary.

raulk reviewed (commented):
Another thing, and this one may lead to more discussion.

**We should not fall back to a full search on miss.**

These guardrails are _not_ safe to have unless we bound the lookup, which we can't do because AFAIK Eth tools expect to be able to access historical receipts.

Note that Eth clients can serve receipts with a single DB lookup, and node operators like Infura, etc. offer instantaneous access to historical receipts.

The receipt for [this transaction](https://etherscan.io/tx/0xfaa6150811c3e9f377fb6872a8cd5a92daa38a98ec7a40f3c848e4c76c460c89) from 7+ years ago was loaded in sub-second response times:

```
⟩ time curl --url https://mainnet.infura.io/v3/<key> \
      -X POST \
      -H "Content-Type: application/json" \
      -d '{"jsonrpc":"2.0","method":"eth_getTransactionReceipt","params":["0xfaa6150811c3e9f377fb6872a8cd5a92daa38a98ec7a40f3c848e4c76c460c89"],"id":1}'
{"jsonrpc":"2.0","id":1,"result":{"blockHash":"0xe33a89a33196a9c6f11a51940823711b18ce10800a5c49fe65f473c75b430965","blockNumber":"0x7a126","contractAddress":null,"cumulativeGasUsed":"0x5208","effectiveGasPrice":"0xba43b7400","from":"0x935e1a56bf95c223e93a41e8bdf63959e99ae080","gasUsed":"0x5208","logs":[],"logsBloom":"0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000","root":"0xc1d169faae783f0a94f4edc1f50f44ab3179a053949eede9d13db674e678c819","to":"0x8263b7757b139904c7cd8511b3d0a6d1479848ba","transactionHash":"0xfaa6150811c3e9f377fb6872a8cd5a92daa38a98ec7a40f3c848e4c76c460c89","transactionIndex":"0x0","type":"0x0"}}
________________________________________________________
Executed in  422.21 millis    fish           external
   usr time   14.40 millis    0.07 millis   14.34 millis
   sys time   11.65 millis    3.83 millis    7.82 millis
```

raulk reviewed (commented):
One more thing: we need to populate the index on snapshot import (added to the TODO list above)

vyzo commented:
Update: points (1), (2), and (5) addressed; will reify and use the events observer next instead of a head change coalescer.

magik6k reviewed (commented):
Looks reasonable, couple of comments.

I don't have enough context to have strong opinions on exact semantics, so I'll leave that to @raulk

vyzo commented on chain/index/msgindex.go:
```diff
@@ -0,0 +1,438 @@
+const (
+	// prepared stmts
+	dbqGetMessageInfo       = "SELECT tipset_cid, epoch FROM messages WHERE cid = ?"
+	dbqInsertMessage        = "INSERT INTO messages VALUES (?, ?, ?)"
```
there should be no conflict, really.

But maybe we do want some policy.

vyzo replied:
ok, the shed tool might manual insert stuff over existing thinges, so we probably want an ON CONFLICT REPLACE in the table definition.

vyzo replied:
https://github.com/filecoin-project/lotus/pull/10452/commits/1676d5148436136c74afc9dcc050106750e5aa42

vyzo commented on chain/index/msgindex.go:
```diff
@@ -0,0 +1,438 @@
+	for _, hc := range pend {
+		for _, ts := range hc.rev {
+			if err := x.doRevert(ctx, tx, ts); err != nil {
+				tx.Rollback() //nolint
```
ack, will do.

vyzo replied:
done.

vyzo commented on node/modules/msgindex.go:
```diff
@@ -0,0 +1,36 @@
+	return msgIndex, nil
+}
+
+func DummyMsgIndex() (index.MsgIndex, error) {
```
ok, will just drop the error :)

vyzo replied:
done.

vyzo commented on node/builder_chain.go:
```diff
@@ -79,6 +80,7 @@ var ChainNode = Options(
 	Override(new(stmgr.Executor), consensus.NewTipSetExecutor(filcns.RewardFunc)),
 	Override(new(consensus.Consensus), filcns.NewFilecoinExpectedConsensus),
 	Override(new(*store.ChainStore), modules.ChainStore),
+	Override(new(index.MsgIndex), modules.MsgIndex),
```
yeah, good point.
Should I add a section in the config with an enabled bool? We might want to add more knobs.

vyzo replied:
https://github.com/filecoin-project/lotus/pull/10452/commits/db8b593c9bb751128893fa2809ea2c0720bbba64

vyzo commented:
@raulk the events observer is not as straightforward to extract, as (1) it has a sprawl of dependencies on other things in the package and (2) it has no tests!

I think we should do this extraction and replacement of the coalescer in a follow up pr.
For now, we can add a check on read and wait until the height exceeds the record height before returning it, effectively implementing confidence.

Note: I do agree that the events observer is more efficient, as we do (potentially a lot) less db i/o. But I don't think it is prudent to do it now.

Opinions? @magik6k @arajasek yours matters too.

vyzo commented on chain/index/interface.go:
```diff
@@ -0,0 +1,43 @@
+
+// MsgIndex is the interface to the message index
+type MsgIndex interface {
+	// GetMsgInfo retrieves the message metadata through the index.
+	GetMsgInfo(ctx context.Context, m cid.Cid) (MsgInfo, error)
```
it's the chain message cid. 

So it is signed for SECP messages, and (if I understand the whole bls thingie correctly) unsigned for BLS? Sigh, this thing drives me nuts.

magik6k replied:
Yeah, bls is unsigned on chain.

vyzo replied:
https://github.com/filecoin-project/lotus/pull/10452/commits/5e011d536b0ff2af2cc2958b9643dc420876e118

vyzo commented:
Actually we don't even need to add the check in the implementation, the client can implement its own confidence by comparing with the head -- this in stmgr.

Also, we can increase the coalescer delay all the way to 30s probably (and override in the unit tests so that they don't last forever.

vyzo commented:
Re: fsync: It looks like we don't have to do anything to get synchronous writes for transactions.
See https://www.sqlite.org/pragma.html#pragma_synchronous

vyzo commented:
Re: multiprocess access to db: https://www.sqlite.org/pragma.html#pragma_locking_mode

Here again it looks like we don't have to do anything.

vyzo commented:
hrm, CI is doing something weird with my itest:
```
# command-line-arguments [command-line-arguments.test]
itests/msgindex_test.go:88:11: undefined: apiSuite
WARN invalid TestEvent: FAIL	command-line-arguments [build failed]
bad output from test2json: FAIL	command-line-arguments [build failed]

=== Errors
itests/msgindex_test.go:88:11: undefined: apiSuite

DONE 0 tests, 1 error in 108.253s

Exited with code exit status 2

```

needless to say, it works fine locally.
@magik6k any ideas here?

magik6k commented on cmd/lotus-shed/msgindex.go:
```diff
@@ -0,0 +1,206 @@
+		}
+		epochs := cctx.Int("epochs")
+
+		dbPath := path.Join(cctx.String("repo"), "sqlite", "msgindex.db")
```
Is `~/.lotus` actually expanded by sql.Open?

vyzo replied:
https://github.com/filecoin-project/lotus/pull/10452/commits/1676d5148436136c74afc9dcc050106750e5aa42

magik6k reviewed (approved):
LGTM, tho @raulk should also have a chance to review before merging.

arajasek commented on chain/index/interface.go:
```diff
@@ -0,0 +1,45 @@
+type MsgInfo struct {
+	// the message this record refers to
+	Message cid.Cid
+	// the tipset where this messages was included
```
```suggestion
	// the tipset where this message was included
```

arajasek commented on chain/stmgr/searchwait.go:
```diff
@@ -159,6 +173,41 @@ func (sm *StateManager) SearchForMessage(ctx context.Context, head *types.TipSet
+		return nil, nil, cid.Undef, err
+	}
+
+	// check that it is indeed the parent of the inclusion tipset
```
This comment is inverted -- the inclusion tipset is the parent of xts, not the other way around.

vyzo replied:
it does read funny, indeed; will fix.

arajasek commented on chain/stmgr/searchwait.go:
```diff
@@ -145,7 +146,20 @@ func (sm *StateManager) SearchForMessage(ctx context.Context, head *types.TipSet
 		return head, r, foundMsg, nil
 	}
 
-	fts, r, foundMsg, err := sm.searchBackForMsg(ctx, head, msg, lookbackLimit, allowReplaced)
+	fts, r, foundMsg, err := sm.searchForIndexedMsg(ctx, mcid, msg)
```
As currently written, this will not benefit calls to `WaitForMessage` (the API's `StateWaitMsg`). You can either perform a quick refactor here to speed up that method too, or just open an enhancement issue?

vyzo replied:
let's TODO it for the follow up.

vyzo replied:
added TODO.

arajasek commented on chain/stmgr/searchwait.go:
```diff
@@ -159,6 +173,41 @@ func (sm *StateManager) SearchForMessage(ctx context.Context, head *types.TipSet
+	}
+
+	r, foundMsg, err := sm.tipsetExecutedMessage(ctx, xts, mcid, m.VMMessage(), false)
+	return xts, r, foundMsg, err
```
This is risky -- it's possible that `tipsetExecutedMessage` "succeeds" (doesn't error), [without actually finding the message](https://github.com/filecoin-project/lotus/blob/47646cbd6966ee440da5ffe7933e73ec67853f39/chain/stmgr/searchwait.go#L333). As written, we `SearchForMessage` will successfully return incomplete / wrong information in that case (see line 153).

We should check that `r` and `foundMsg` are non-nil / non-undef either here, or in callers of this (or both).

vyzo replied:
ok

vyzo replied:
added check in call site.

arajasek reviewed (commented):
Thanks for the PR! This generally looks good. I have a few open questions:

- Did you try / can we try having the ChainStore own this index? I think it belongs there more, "searching back for a message" is very much a chain operation (even if its caller is currently the stmgr).
- As written, we can only backfill manually. That means if your node is down for 30 mins and then rebooted, you'll be missing index entries for that period. Are we okay with that?
- We currently only reconcile on startup. Did you consider triggering the reconciliation process internally if the indexer ends up in some wonky state (eg. errors while processing head change). It also might be useful for callers of the index to be able to "tell" the indexer that it's inconsistent, and trigger a reconciliation (eg. if for some reason we didn't _actually_ find the message in the returned tipset)
- On the flip-side, as written with the current user of this index, we _only_ need the Tipset CID in the MsgInfo struct for reconciliation, a process we don't expect to need -- we don't actually need it for the lookback, because we load the "execution" tipset and pull the info out of it anyway. This means that we could drop the size of this db by ~half (one of the 2 CIDs), and have this just me a `(msgCid, height)` mapping. In essence we can get a much smaller database footprint (50%) at the cost of much slower / worse reconciliation (either manually inspecting every record, or just dropping any records that callers report as "bad").
- I have pretty high confidence in the correctness of this thanks to how simple it is. Even so, could we consider adding a "paranoid" mode, that uses the cache, then does the lookback anyway, and screams if results don't match? This can be run by us, as well as any integration partners with high security standards for this portion of the codebase (we have some such partners).
- I think it'll be the case that if this is slow, it'll slow down overall sync. Is that so, and if so can we maybe add some timing metrics?
- I'm strongly in favour of having a single table with the minor duplication here :+1:

Stebalien commented on chain/index/msgindex.go:
```diff
@@ -0,0 +1,458 @@
+		return xerrors.Errorf("error preparing statement: %w", err)
+	}
+
+	curTs := cs.GetHeaviestTipSet()
```
This is racy unless we can guarantee that the chainstore hasn't been "started" yet. Is that the case @arajasek?

Otherwise, we need a way to atomically get the head and subscribe at the same time. Alternatively, we can defer this reconcile step until we get the first event from the chainstore and reconcile then.

Stebalien replied:
Actually, that second part is kind of nice because it avoids blocking on start.

vyzo replied:
Yeah, but I am not convinced it is a problem -- let's wait for @arajasek to chime in here.

arajasek replied:
Yeah, I don't think this is a problem -- the chainstore shouldn't have "started", and even if it had, I don't think I see a problem if this races?

I do agree with @Stebalien that reconciling on first event would be nicer, though, but it's non-blocking IMO.

Stebalien commented on chain/index/msgindex.go:
```diff
@@ -0,0 +1,458 @@
+	msgIndex := &msgIndex{
+		db:     db,
+		cs:     cs,
+		sema:   make(chan struct{}, 1),
```
ultra-nit: this is a flag, not a semaphore.

vyzo replied:
ha! I can rename if you insist, you are technically right.

vyzo commented:
@arajasek thanks for the review, these are some great questions.

> Did you try / can we try having the ChainStore own this index? I think it belongs  there more, "searching back for a message" is very much a chain operation (even if its caller is currently the stmgr).

I have considered this and I think it has _great_ potential. 
If we had the index owned by the ChainStore a whole lot of things could be significantly improved in terms of performance.

Some very basic things like `GetTipsetByHeight`, `GetMessagesForTipset`, etc can go through the index.

The main reason  I didn't do it for the MVP is that it is a much more invasive change in the most consensus critical part of the system. So a different standard of applicability and testing must be adhered there, and I'd rather do it in a follow up.

The main question is whether it needs to become synchronous in terms of head change notification processing in order to make ChainStore keep the authoritative source of truth invariant. We discussed this with @Stebalien and we think it doesn't have to, we can check if the current tipset (or the one we are based on in general) is in the index, and in that case proceed and use it -- otherwise fallback to what we already have. We should probably tweak down the coalescer delay a bit too, and generally experiment with it. Note that the reason for not wanting synchronous operation is that we will thrash the database with writes/deletes while the tipset is building and this is something we want to avoid for obvious reasons (perf, durability of SSDs, and so on).

At any rate, I think that having the index owned by the ChainStore is a natural progression from the MVP and a very important follow up for subsequent prs.

> As written, we can only backfill manually. That means if your node is down for 30 mins and then rebooted, you'll be missing index entries for that period. Are we okay with that?

I think we are ok for MVP, but we do need to do something about it. The first thing to do (as discussed with @raulk) is to fill the index on snapshot import, and that should probably cover us. We can also consider backfilling when first creating the index as well, but I think snapshot import is a better place to handle this. 

At any rate, part of the follow up plan to address this.

> We currently only reconcile on startup. Did you consider triggering the reconciliation process internally if the indexer ends up in some wonky state (eg. errors while processing head change). It also might be useful for callers of the index to be able to "tell" the indexer that it's inconsistent, and trigger a reconciliation (eg. if for some reason we didn't actually find the message in the returned tipset)

Yeah, that's a good point. The problem is the synchronous nature of it, hence startup seemed like the logical place to do it. We should definitely consider expanding the adaptive capabilities in follow up, but I am not convinced it is necessary.

> On the flip-side, as written with the current user of this index, we only need the Tipset CID in the MsgInfo struct for reconciliation, a process we don't expect to need -- we don't actually need it for the lookback, because we load the "execution" tipset and pull the info out of it anyway. This means that we could drop the size of this db by ~half (one of the 2 CIDs), and have this just me a (msgCid, height) mapping. In essence we can get a much smaller database footprint (50%) at the cost of much slower / worse reconciliation (either manually inspecting every record, or just dropping any records that callers report as "bad").

I don't think we should do that, having the tipset cid in the index opens up great opportunities. See the notes above about ChainStore ownership, but at the very least we can return the execution tipset this way in addition to the inclsion tipset on queries if we have it. I left a TODO for an optimization follow up in the search implementation.

> I have pretty high confidence in the correctness of this thanks to how simple it is. Even so, could we consider adding a "paranoid" mode, that uses the cache, then does the lookback anyway, and screams if results don't match? This can be run by us, as well as any integration partners with high security standards for this portion of the codebase (we have some such partners).

That's probably a good idea, and it will definitely helps us build confidence harder.

> I think it'll be the case that if this is slow, it'll slow down overall sync. Is that so, and if so can we maybe add some timing metrics?

I don't think it will be slow (sqlite is an amazing piece of software), but you are absolutely right that we should have some timing measurements. I will add some logging to that regard.

> I'm strongly in favour of having a single table with the minor duplication here +1

yeah, it makes things so much simpler! And it is not that big of a waste, that 2x is not that bad.

vyzo commented on chain/stmgr/searchwait.go:
```diff
@@ -159,6 +173,41 @@ func (sm *StateManager) SearchForMessage(ctx context.Context, head *types.TipSet
+func (sm *StateManager) searchForIndexedMsg(ctx context.Context, mcid cid.Cid, m types.ChainMsg) (*types.TipSet, *types.MessageReceipt, cid.Cid, error) {
+	minfo, err := sm.msgIndex.GetMsgInfo(ctx, mcid)
+	if err != nil {
+		return nil, nil, cid.Undef, err
```
ack, will do.

vyzo replied:
done.

vyzo commented on chain/index/msgindex.go:
```diff
@@ -0,0 +1,458 @@
+
+	db, err := sql.Open("sqlite3", dbPath)
+	if err != nil {
+		// TODO [nice to have]: automaticaly delete corrupt databases
```
yes, of course. 

Part of making the TODO was to discuss them, see what we need to do in the initial MVP, and the rest can go into a follow up issue.

vyzo commented on chain/index/msgindex.go:
```diff
@@ -0,0 +1,458 @@
+	// do it in the background to avoid blocking head change processing
+	x.mx.Lock()
+	x.pend = append(x.pend, headChange{rev: rev, app: app})
+	// TODO log loudly if this is building backlog (it shouldn't but better be safe on this)
```
ok.

vyzo commented on chain/index/msgindex.go:
```diff
@@ -0,0 +1,458 @@
+		case <-x.sema:
+			err := x.processHeadChanges(ctx)
+			if err != nil {
+				// TODO should we shut down the index altogether? we just log for now.
```
ok, it is a reasonable thing to do.

vyzo replied:
done.

vyzo commented on chain/index/msgindex.go:
```diff
@@ -0,0 +1,458 @@
+	for _, hc := range pend {
+		for _, ts := range hc.rev {
+			if err := x.doRevert(ctx, tx, ts); err != nil {
+				if err := tx.Rollback(); err != nil {
```
ha, ok.

vyzo commented on chain/index/msgindex.go:
```diff
@@ -0,0 +1,458 @@
+	//  1. Count messages in index; if none, trivially reconciled.
+	//     TODO we may consider populating the index in that case
+	//  2. Find the minimum tipset in the index; this will mark the end of the reconciliation walk
+	//  3. Walk from current tipset until we find a tipset in the index.
```
Do the backfill? Probably not that much slower, but it is something we can measure and implement in the follow up.

I went for KISS here, that's the only reason i didn't backfill while reconciling.

vyzo commented on chain/index/interface.go:
```diff
@@ -0,0 +1,45 @@
+	Message cid.Cid
+	// the tipset where this messages was included
+	TipSet cid.Cid
+	// the epoch whre this message was included
```
done.

vyzo commented:
Updates:
- lotus-shed now expands the repo dir to avoid the tilde footgun
- added an ON CONFLICT clause in the table for the message cid, to avoid conflicts from stuff inserted manually with shed
- addressed @arajasek review comments.

vyzo commented:
yessir, i shall do.

jennijuju commented:
merge block: can we have some user documentation is the PR description? i.e: what kind of the configuration exists? any tooling to manage the db?

vyzo commented:
> merge block: can we have some user documentation is the PR description? i.e: what kind of the configuration exists? any tooling to manage the db? 

@jennijuju done, updated the pr description.

vyzo commented:
@arajasek follow up issue: https://github.com/filecoin-project/lotus/issues/10536
